### PR TITLE
Update omnibus package scripts

### PR DESCRIPTION
**Pull request by julien-lebot**
### What does this PR do?

Update the IoT Agent omnibus' package-scripts to support more platforms.

### Motivation

The original script bailed when the installation system was not a Debian-based system.

### Describe your test plan

1) IoT Agent installs on a Debian-based system
- Install the IoT Agent on [Ubuntu/Debian]
  - The install succeeds
  - The Agent is functional (check with `sudo datadog-agent status`)
  - The `dd-agent` user and `dd-agent` group exists
- Uninstall the Agent
  a) With `--purge`
  b) Without `--purge`
  In both cases the uninstall succeeds. With `--purge` no trace of the Agent left in `/etc/datadog-agent`, `/opt/datadog-agent` or `/var/log/datadog`. There is no trace of the `dd-agent` user, nor `dd-agent` group.

2) IoT Agent installs on a system using RPM
- Install the IoT Agent on [CentOS/RedHat]
  - The install succeeds
  - The Agent is functional (check with `sudo datadog-agent status`)
  - The `dd-agent` user and `dd-agent` group exists
- Uninstall the Agent
  - The uninstall succeeds
  - The `dd-agent` user and `dd-agent` group still exists
  - The `/etc/datadog-agent`, `/opt/datadog-agent` or `/var/log/datadog` are still there.

### CODE REVIEW

File: omnibus/package-scripts/iot-agent/preinst [hunk 1]
```diff
@@ -20,6 +25,12 @@ if [ "$DISTRIBUTION" != "Darwin" ]; then
...
         elif command -v initctl >/dev/null 2>&1; then
             initctl stop $SERVICE_NAME || true
+        elif [ "$DISTRIBUTION_FAMILY" = "Debian" ]; then
+            if command -v service >/dev/null 2>&1; then
+                service $SERVICE_NAME stop || true
+            else
+                echo "[ WARNING ]\tCannot detect a supported init system. The datadog-agent package only provides service files for systemd, upstart and sysvinit."
+            fi
         else
             echo "[ WARNING ]\tCannot detect a supported init system. The datadog-agent package only provides service files for systemd and upstart."
```
Comment (albertvaka) [inline omnibus/package-scripts/iot-agent/preinst:35]:
```suggestion
            echo "[ WARNING ]\tCannot detect a supported init system. The datadog-iot-agent package only provides service files for systemd and upstart."
```

File: omnibus/package-scripts/iot-agent/preinst [hunk 2]
```diff
@@ -20,6 +25,12 @@ if [ "$DISTRIBUTION" != "Darwin" ]; then
             systemctl stop $SERVICE_NAME || true
         elif command -v initctl >/dev/null 2>&1; then
             initctl stop $SERVICE_NAME || true
+        elif [ "$DISTRIBUTION_FAMILY" = "Debian" ]; then
+            if command -v service >/dev/null 2>&1; then
+                service $SERVICE_NAME stop || true
+            else
+                echo "[ WARNING ]\tCannot detect a supported init system. The datadog-agent package only provides service files for systemd, upstart and sysvinit."
```
Comment (albertvaka) [inline omnibus/package-scripts/iot-agent/preinst:32]:
```suggestion
                echo "[ WARNING ]\tCannot detect a supported init system. The datadog-iot-agent package only provides service files for systemd, upstart and sysvinit."
```

File: omnibus/package-scripts/iot-agent/postinst [hunk 3]
```diff
@@ -14,79 +16,91 @@ DISTRIBUTION=$(lsb_release -d 2>/dev/null | grep -Eo $KNOWN_DISTRIBUTION || grep
...
+        if command -v systemctl >/dev/null 2>&1; then
+            # Force systemd to ignore the sysvinit scripts. Only cosmetic, remove some irrelevant warnings during upgrade
+            SYSTEMCTL_SKIP_SYSV=true systemctl enable $SERVICE_NAME || echo "[ WARNING ]\tCannot enable $SERVICE_NAME with systemctl"
+        elif command -v initctl >/dev/null 2>&1; then
+            # Nothing to do, this is defined directly in the upstart job file
+            :
+        elif command -v update-rc.d >/dev/null 2>&1; then
+            update-rc.d $SERVICE_NAME defaults || echo "[ WARNING ]\tCannot enable $SERVICE_NAME with update-rc.d"
+        else
+            echo "[ WARNING ]\tCannot detect a supported init system. The datadog-agent package only provides service files for systemd, upstart and sysvinit."
```
Comment (albertvaka) [inline omnibus/package-scripts/iot-agent/postinst:79]:
```suggestion
            echo "[ WARNING ]\tCannot detect a supported init system. The datadog-iot-agent package only provides service files for systemd, upstart and sysvinit."
```

File: omnibus/package-scripts/iot-agent/postinst [hunk 4]
```diff
@@ -14,79 +16,91 @@ DISTRIBUTION=$(lsb_release -d 2>/dev/null | grep -Eo $KNOWN_DISTRIBUTION || grep
...
+        if [ -f "$CONFIG_DIR/datadog.yaml" ]; then
+            echo "(Re)starting $SERVICE_NAME now..."
+            if command -v systemctl >/dev/null 2>&1; then
+                systemctl restart $SERVICE_NAME || true
+            elif command -v initctl >/dev/null 2>&1; then
+                initctl start $SERVICE_NAME || initctl restart $SERVICE_NAME || true
+            elif command -v service >/dev/null 2>&1; then
+                service $SERVICE_NAME restart || true
+            else
+                echo "[ WARNING ]\tCannot detect a supported init system. The datadog-agent package only provides service files for systemd, upstart and sysvinit."
```
Comment (albertvaka) [inline omnibus/package-scripts/iot-agent/postinst:93]:
```suggestion
                echo "[ WARNING ]\tCannot detect a supported init system. The datadog-iot-agent package only provides service files for systemd, upstart and sysvinit."
```

File: omnibus/package-scripts/iot-agent/posttrans [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    # start/stop policy is already defined in the upstart job file
+    :
+elif [ "$SUSE_SYSVINIT_SUPPORT" = "yes" ]; then
+    if command -v update-rc.d >/dev/null 2>&1; then
+        update-rc.d $SERVICE_NAME defaults || echo "[ WARNING ]\tCannot enable $SERVICE_NAME with update-rc.d"
+    elif command -v chkconfig >/dev/null 2>&1; then
+        chkconfig --set $SERVICE_NAME on || echo "[ WARNING ]\tCannot enable $SERVICE_NAME with chkconfig"
+    fi
+else
+    echo "[ WARNING ]\tCannot detect a supported init system. The datadog-agent package only provides service files for systemd and upstart."
```
Comment (albertvaka) [inline omnibus/package-scripts/iot-agent/posttrans:49]:
```suggestion
    echo "[ WARNING ]\tCannot detect a supported init system. The datadog-iot-agent package only provides service files for systemd and upstart."
```

File: omnibus/package-scripts/iot-agent/posttrans [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+if [ -f "$CONFIG_DIR/datadog.yaml" ]; then
+    echo "(Re)starting $SERVICE_NAME now..."
+    if command -v systemctl >/dev/null 2>&1; then
+        systemctl restart $SERVICE_NAME || true
+    elif command -v initctl >/dev/null 2>&1; then
+        initctl start $SERVICE_NAME || initctl restart $SERVICE_NAME || true
+    elif [ "$SUSE_SYSVINIT_SUPPORT" = "yes" ] && command -v service >/dev/null 2>&1; then
+        service $SERVICE_NAME restart || true
+    else
+        echo "[ WARNING ]\tCannot detect a supported init system. The datadog-agent package only provides service files for systemd and upstart."
```
Comment (albertvaka) [inline omnibus/package-scripts/iot-agent/posttrans:63]:
```suggestion
        echo "[ WARNING ]\tCannot detect a supported init system. The datadog-iot-agent package only provides service files for systemd and upstart."
```

**Review by albertvaka — APPROVED** [review]

**Review by albertvaka — DISMISSED** [review]
LGTM. An improvement, given the two agent flavors use almost identical scripts, would be to reuse code between them. I'm not sure how easy it is, though, so let's go with this for now. The same thing already happens with the Windows installer XMLs.

**Comment by albertvaka** [comment]
Tested on:
- [x] CentOS 8.
- [ ] CentOS 6.
- [ ] Ubuntu 14.04
- [ ] Ubuntu 18.04

File: omnibus/package-scripts/iot-agent/prerm [hunk 7]
```diff
@@ -36,6 +36,11 @@ deregister_agent()
     fi
 }
 
+if [ "$DISTRIBUTION" = "Darwin" ]; then
+	echo "The IoT Agent is not supported on MacOS"
+	exit 0
+fi
+
```
Comment (KSerrania) [inline omnibus/package-scripts/iot-agent/prerm:43]: I think there's no need to add anything for MacOS in the `prerm`, it's not shipped in the `.pkg` (only the `preinst` and `postinst` are).
See: https://github.com/DataDog/omnibus-ruby/blob/94d7bdf36ca2b4adb8167a3d54bb30732fefce47/lib/omnibus/packagers/pkg.rb#L20-L27

File: omnibus/package-scripts/iot-agent/postinst (hunk 3 above)
Comment (KSerrania) [reply omnibus/package-scripts/iot-agent/postinst:79]: Are we really supporting `sysvinit`? I only see `upstart` & `systemd` mentioned in the project file: https://github.com/DataDog/datadog-agent/blob/29fbbea6d6dc44dcd4f347331cedbf48743977f1/omnibus/config/projects/iot-agent.rb#L76-L95

File: omnibus/package-scripts/iot-agent/postinst [hunk 8]
```diff
@@ -14,79 +16,91 @@ DISTRIBUTION=$(lsb_release -d 2>/dev/null | grep -Eo $KNOWN_DISTRIBUTION || grep
...
+        if [ -f "$CONFIG_DIR/datadog.yaml" ]; then
+            echo "(Re)starting $SERVICE_NAME now..."
+            if command -v systemctl >/dev/null 2>&1; then
+                systemctl restart $SERVICE_NAME || true
+            elif command -v initctl >/dev/null 2>&1; then
+                initctl start $SERVICE_NAME || initctl restart $SERVICE_NAME || true
+            elif command -v service >/dev/null 2>&1; then
+                service $SERVICE_NAME restart || true
+            else
+                echo "[ WARNING ]\tCannot detect a supported init system. The datadog-iot-agent package only provides service files for systemd, upstart and sysvinit."
```
Comment (KSerrania) [inline omnibus/package-scripts/iot-agent/postinst:93]: Same here: do we support `sysvinit` for the `iot-agent` package?

Comment (albertvaka) [reply omnibus/package-scripts/iot-agent/postinst:93]: We don't, it's probably an oversight from copy-pasting.

File: omnibus/package-scripts/iot-agent/postinst [hunk 9]
```diff
@@ -1,8 +1,10 @@
 #!/bin/sh
 #
 # Perform necessary datadog-agent setup steps after package is installed.
+# NOTE: for .rpm, see posttrans instead
 #
-# .deb: STEP 5 of 5
+# .deb: STEP 2 of 5
```
Comment (mx-psi) [inline omnibus/package-scripts/iot-agent/postinst:6]: Shouldn't this actually be step 5 of 5?

**Review by mx-psi — APPROVED** [review]
LGTM!

**Pull request merged**